tests/conftest.py

```diff
@@ -55,9 +56,118 @@ def pytest_addoption(parser):
     )
 
 
+def _modify_compose_for_testing(request):
+    demo = "../docker-compose.demo.yml"
+    testing = "../docker-compose.testing.yml"
+    storage_proxy_demo = "../storage-proxy/docker-compose.storage-proxy.demo.yml"
+    storage_proxy_testing = "../storage-proxy/docker-compose.storage-proxy.testing.yml"
+
+    # Published ports in docker-compose-demo and docker compose.storage-proxy.demo
+    demo_ports = ["9000:9000", "8080:8080", "443:443", "80:80"]
+    demo_storage_proxy_ports = ["9000:9000"]
+
+    # Remove published ports
+    for ports, out, file in zip(
+        [demo_ports, demo_storage_proxy_ports],
+        [testing, storage_proxy_testing],
+        [demo, storage_proxy_demo],
+    ):
+
+        command = ["sed"]
+        for port in ports:
+            command.extend(["-e", f"/{port}/d"])
+        subprocess.run(command + ["-e", "/ports:/d", file], stdout=open(out, "w"))
+
+    # Disable download speed limits
+    subprocess.run(["sed", "-e", "s/DOWNLOAD_SPEED/#DOWNLOAD_SPEED/", "-i", testing])
+
+    # Whitelist all IPs/DNS names in the gateway
+    subprocess.run(
+        ["sed", "-e", "s/ALLOWED_HOSTS: .*/ALLOWED_HOSTS: ~./", "-i", testing]
+    )
+
+
+@pytest.fixture(scope="session", autouse=True)
+def modify_compose_for_testing(request):
+    return _modify_compose_for_testing(request)
+
+
+def _extract_fs_from_image(request, client_compose_file):
+    d = tempfile.TemporaryDirectory()
+
+    def cleanup():
+        shutil.rmtree(d.name, ignore_errors=True)
+
+    request.addfinalizer(cleanup)
+
+    image_type = "mender-client"
+    if "gateway" in client_compose_file:
+        image_type = "mender-gateway"
+
+    with open(client_compose_file) as f:
+        data = yaml.safe_load(f)
+        image = data["services"][image_type]["image"]
+
+    subprocess.check_call(
+        [
+            "docker",
+            "run",
+            "--rm",
+            "--privileged",
+            "--entrypoint",
+            "/extract_fs",
+            "--volume",
+            d.name + ":/output",
+            image,
+        ]
+    )
+
+    return d.name
+
+
 @pytest.fixture(scope="session")
 def valid_image(request):
-    return "core-image-full-cmdline-%s.ext4" % machine_name
+    dirname = _extract_fs_from_image(
+        request, os.path.join(THIS_DIR, "..", "docker-compose.client.yml")
+    )
+    return dirname, f"core-image-full-cmdline-{machine_name}.ext4"
```

Comment:
Q: Should this be a string concatenation, and not returning a two-tuple? :thinking: